File: classical-problems/leetcode50-Pow/Solution.py
class Solution:
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        p = n
        if p < 0: p = -p
        ans = 1.
        while p:
            if p & 1: ans *= x
            x *= x
            p >>= 1
        if n < 0: return 1 / ans
        return ans

File: classical-problems/leetcode50-Pow/Solution.java
class Solution {
    public double myPow(double x, int n) {
        long p = 1l * n;
        if (p < 0) p = -p;
        double ans = 1.;
        while (p != 0) {
            if ((p & 1) != 0) ans *= x;
            x *= x;
            p >>= 1;
        }
        return n < 0 ? 1 / ans : ans;
    }
}
